### PR TITLE
Added pj_turn_sock_connect for TURN TCP Allocations

### PR DESCRIPTION
**Pull request by xdyyll-aes256**
Hey. I found that pj_turn_sock couldn't initiate a connect request, so I added this feature to implement RFC 6062:

`pj_turn_sock_connect(turn_sock, &turn_sess_info.relay_addr, pj_sockaddr_get_len(&turn_sess_info.relay_addr));`

disconnect alsa has added

DEMO:
```
#include <pjnath.h>
#include <pjlib-util.h>
#include <pjlib.h>

#define SERVER_IP	"194.5.78.83"
#define SERVER_PORT	3478

pj_caching_pool		cp;
pj_pool_t		*pool;

pj_ice_strans_cfg	ice_cfg;
pj_stun_auth_cred 	cred;

pj_thread_t     	*thread;
pj_bool_t 		running = PJ_TRUE;


static int ice_worker_thread(void *unused)
{
    PJ_UNUSED_ARG(unused);
    while (running) {
        pj_time_val delay = {0, 500};
        pj_timer_heap_poll(ice_cfg.stun_cfg.timer_heap, NULL);
        pj_ioqueue_poll(ice_cfg.stun_cfg.ioqueue, &delay);
    }
}

void turn_on_rx_data(pj_turn_sock *turn_sock,
		     void *pkt,
		     unsigned pkt_len,
		     const pj_sockaddr_t *peer_addr,
		     unsigned addr_len)
{
    printf("turn_on_rx_data %s\n", (char *)pkt);
}

pj_bool_t turn_on_data_sent(pj_turn_sock *sock, pj_ssize_t sent)
{
    printf("on_data_sent\n");
}

void turn_on_state(pj_turn_sock *turn_sock,
		   pj_turn_state_t old_state,
		   pj_turn_state_t new_state)
{
    printf("statue: %s => %s\n",
	   pj_turn_state_name(old_state),
	   pj_turn_state_name(new_state));
}

pj_status_t turn_on_connection_attempt(pj_turn_sock *turn_sock,
				       pj_uint32_t conn_id,
				       const pj_sockaddr_t *peer_addr,
				       unsigned addr_len)
{
    printf("on_connection_attempt\n");
    return PJ_SUCCESS;
}

void turn_on_connection_status(pj_turn_sock *turn_sock,
			       pj_status_t status,
			       pj_uint32_t conn_id,
			       const pj_sockaddr_t *peer_addr,
			       unsigned addr_len)
{
    printf("on_connection_status: %d\n", status);
}

int start_turn(int argc, char *argv[]) {
    pj_str_t server_ip = pj_str(SERVER_IP);
    pj_uint16_t server_port = SERVER_PORT;

    pj_caching_pool_init(&cp, &pj_pool_factory_default_policy, 0);
    pool = pj_pool_create(&cp.factory, "main", 10000, 10000, NULL);
    if (pool == NULL) {
        PJ_PERROR(1, (__FILE__, PJ_ENOMEM, "Error createing poll"));
        return PJ_ENOMEM;
    }

    pj_ice_strans_cfg_default(&ice_cfg);
    ice_cfg.stun_cfg.pf = &cp.factory;
    pj_timer_heap_create(pool, 1000, &ice_cfg.stun_cfg.timer_heap);
    pj_ioqueue_create(pool, 16, &ice_cfg.stun_cfg.ioqueue);

    pj_thread_create(pool, "ice", &ice_worker_thread, NULL, 0, 0, &thread);

    pj_status_t status;
    pj_turn_sock *turn_sock, *turn_sock2;
    pj_turn_sock_cb turn_sock_cb;

    turn_sock_cb.on_rx_data = turn_on_rx_data;
    turn_sock_cb.on_data_sent = turn_on_data_sent;
    turn_sock_cb.on_state = turn_on_state;
    turn_sock_cb.on_connection_attempt = turn_on_connection_attempt;
    turn_sock_cb.on_connection_status = turn_on_connection_status;

    cred.type = PJ_STUN_AUTH_CRED_STATIC;
    cred.data.static_cred.username = pj_str("juice_test");
    cred.data.static_cred.data_type = PJ_STUN_PASSWD_PLAIN;
    cred.data.static_cred.data = pj_str("28245150316902");

    pj_turn_alloc_param alloc_param;
    pj_turn_alloc_param_default(&alloc_param);
    alloc_param.peer_conn_type = PJ_TURN_TP_TCP;

    status = pj_turn_sock_create(&ice_cfg.stun_cfg,
    				 pj_AF_INET(),
				 PJ_TURN_TP_TCP,
				 &turn_sock_cb,
				 NULL,
				 NULL,
				 &turn_sock);

    status = pj_turn_sock_create(&ice_cfg.stun_cfg,
    			         pj_AF_INET(),
				 PJ_TURN_TP_TCP,
				 &turn_sock_cb,
				 NULL,
				 NULL,
				 &turn_sock2);

    status = pj_turn_sock_alloc(turn_sock,
    				&server_ip,
				server_port,
				NULL,
				&cred,
				&alloc_param);

    status = pj_turn_sock_alloc(turn_sock2,
    				&server_ip,
				server_port,
				NULL,
				&cred,
				&alloc_param);

    pj_thread_sleep(5000);

    pj_sockaddr addr[8], addr2[8];
    pj_turn_session_info turn_sess_info;
    pj_turn_session_info turn_sess_info2;

    int addr_cnt = 0;
    status = pj_turn_sock_get_info(turn_sock, &turn_sess_info);
    addr[addr_cnt++] = turn_sess_info.relay_addr;

    int addr_cnt2 = 0;
    status = pj_turn_sock_get_info(turn_sock2, &turn_sess_info2);
    addr2[addr_cnt2++] = turn_sess_info.server;

    /*
     * Need to setup permission otherwise
     * ConnnectBind request to turn_sock2 will be failed
     */
    pj_turn_sock_set_perm(turn_sock2, addr_cnt, addr, 0);
    pj_thread_sleep(2000);
    pj_turn_sock_connect(turn_sock,
			 &turn_sess_info2.relay_addr,
			 pj_sockaddr_get_len(&turn_sess_info2.relay_addr));

    pj_thread_sleep(5000);

    int i;
    for (i = 0; i < 5; i++) {
	pj_turn_sock_sendto(turn_sock, "hello", 6,
			    &turn_sess_info2.relay_addr,
			    sizeof(turn_sess_info2.relay_addr));
	pj_thread_sleep(1000);
    }

    pj_turn_sock_disconnect(turn_sock,
			    &turn_sess_info2.relay_addr,
			    pj_sockaddr_get_len(&turn_sess_info2.relay_addr));

    pj_turn_sock_destroy(turn_sock);

    pj_ioqueue_destroy(ice_cfg.stun_cfg.ioqueue);
    pj_timer_heap_destroy(ice_cfg.stun_cfg.timer_heap);
    pj_pool_release(pool);
    pj_caching_pool_destroy(&cp);

    running = PJ_FALSE;
        
    return 0;
}

int main(int argc, char *argv[]) {
    /*pj_log_set_level(0);*/

    pj_init();
    pjlib_util_init();
    pjnath_init();

    pj_run_app(start_turn, argc, argv, 0);
    pj_shutdown();

    return 0;
}
```

### CODE REVIEW

File: pjnath/include/pjnath/turn_sock.h [hunk 1]
```diff
@@ -622,6 +622,19 @@ PJ_DECL(pj_status_t) pj_turn_sock_sendto(pj_turn_sock *turn_sock,
 PJ_DECL(pj_status_t) pj_turn_sock_bind_channel(pj_turn_sock *turn_sock,
 					       const pj_sockaddr_t *peer,
 					       unsigned addr_len);
+/**
+ * Send Connect request for the specified a peer address.
```
Comment (nanangizz) [inline pjnath/include/pjnath/turn_sock.h:626]: Better to add more info, similar to `pj_turn_session_connect()`.

File: pjnath/src/pjnath/turn_sock.c [hunk 2]
```diff
@@ -1766,3 +1785,196 @@ static void turn_on_connection_bind_status(pj_turn_session *sess,
...
+    PJ_LOG(5,(turn_sock->pool->obj_name, "Trying to connect to peer %s",
+	      pj_sockaddr_print(peer_addr, addrtxt, sizeof(addrtxt), 3)));
+
+    if (turn_sock == NULL) {
+	/* We've been destroyed */
+	return;
+    }
+
+    pj_grp_lock_acquire(turn_sock->grp_lock);
+
```
Comment (nanangizz) [inline pjnath/src/pjnath/turn_sock.c:1820]: Around here we should check & notify app if the connection attempt fails (i.e: `status != PJ_SUCCESS`) via `on_connection_status`.

**Review by nanangizz — COMMENTED** [review]
Yes, currently only TCP accept mode is supported (#2197). Thank you for contributing the patch for TCP connect mode. Could you please sign the CLA.

FYI, we use editor settings as follow: tab=8 chars & indent=4 chars, looks like indentations are a bit off in some places. Also, for consistency, max cols (especially for lib code) should be 80.

**Comment by sauwming** [comment]
This PR should be a great addition to the feature that's trying to be supported in PR #2363 (RFC 6544: TCP Candidates with ICE), right?

If so, please have a look at that PR as well, to avoid any potential conflict when the two are merged together.

**Comment by nanangizz** [comment]
> This PR should be a great addition to the feature that's trying to be supported in PR #2363 (RFC 6544: TCP Candidates with ICE), right?

Yes, this PR is about TURN TCP (connect mode, we have accept mode already) but still nothing to do with ICE yet, and that PR seems to integrate TCP (not sure if it covers both direct TCP and TURN TCP) into ICE.

> 
> If so, please have a look at that PR as well, to avoid any potential conflict when the two are merged together.

Sure. The patch looks huge, also the author mentioned about dirty/hacky, so I guess it requires quite big effort to review+merge it. Anyway, let's put it in the next roadmap meeting.

Conflicts are quite expected actually, e.g: rather major modifications in trickle ICE.

**Comment by xdyyll-aes256** [comment]
@nanangizz thx to your review, I will improve my code quality this weekend.

**Comment by xdyyll-aes256** [comment]
It seems work. The client need to request permission before Connect request:

```
    /*
     * Need to setup permission otherwise
     * ConnnectBind request to turn_sock2 will be failed
     */
    pj_turn_sock_set_perm(turn_sock2, addr_cnt, addr, 0);
    pj_turn_sock_connect(turn_sock,
			 &turn_sess_info2.relay_addr,
			 pj_sockaddr_get_len(&turn_sess_info.relay_addr));
```

**Review by nanangizz — COMMENTED** [review]
Just found few things, some are minors actually.
Thank you.

**Review by xdyyll-aes256 — COMMENTED** [review]
> 
> 
> Just found few things, some are minors actually.
> Thank you.

fixed, thx.

**Review by nanangizz — APPROVED** [review]
Thank you.

**Review by sauwming — APPROVED** [review]

**Pull request merged**